internal/pkg/service/common/etcdop/op/txn.go

```diff
@@ -155,17 +164,29 @@ func (v *TxnOp[R]) lowLevelTxn(ctx context.Context) (*lowLevelTxn[R], error) {
 	out.ifs = make([]etcd.Cmp, len(v.ifs))
 	copy(out.ifs, v.ifs)
 
-	// Map THEN operations
+	// Map Then and ThenTxn operations
 	for i, op := range v.thenOps {
 		// Create low-level operation
-		if lowLevel, err := op.Op(ctx); err == nil {
+		if lowLevel, err := op.Op(ctx); err != nil {
+			errs.Append(errors.PrefixErrorf(err, "cannot create operation [then][%d]", i))
+		} else if lowLevel.Op.IsTxn() {
+			errs.Append(errors.Errorf(`cannot create operation [then][%d]: operation "%T" is a transaction, use ThenTxn, not Then`, i, lowLevel.Op))
+		} else {
 			out.addThen(lowLevel.Op, lowLevel.MapResponse)
+		}
+	}
+	for i, op := range v.thenTxnOps {
+		// Create low-level operation
+		if lowLevel, err := op.Op(ctx); err != nil {
+			errs.Append(errors.PrefixErrorf(err, "cannot create operation [thenTxn][%d]", i))
+		} else if !lowLevel.Op.IsTxn() {
+			errs.Append(errors.Errorf(`cannot create operation [thenTxn][%d]: operation "%T" is not a transaction, use Then, not ThenTxn`, i, lowLevel.Op))
 		} else {
```

Comment:
`Then/ThenTxn` checks are implemented in convertion to a low-level op.
  - Rather, we do not know whether it is a transaction, our `TxnOp` high-level operation may not be used.